docs: add Ford-Fulkerson algorithm topic outline (#2260) (#2730)
Techtonica/curriculum: algorithms/ford-fulkerson.md

## Ford–Fulkerson Algorithm (Maximum Flow)

> 

Reviewer: You are missing some sections from the [topic outline template sections](https://github.com/Techtonica/curriculum/blob/main/_templates/topic-outline.md): 
- Prerequisite Topics
- Motivation
- Objectives
- Specific Things to Learn: Core concepts and skills
- Materials: Links to articles, videos, tool documentation, and pipeline diagrams
- Lesson: Main instructional content covering the "why" and "how", with visual explanations
- Common Mistakes & Misconceptions: Address typical misunderstandings 
- Guided Practice
- Independent Practice: Hands-on exercises where learners build or interact with basic scenarios
- Check for Understanding: Assessment questions verifying understanding of concepts
- Supplemental Materials: Optional resources for deeper learning (advanced topics, tool-specific guides)

Alternatively, please **_do not_** include the following:
- we do not refer to learners or participants as "students"
- refrain from including any instructions that are relevant to maintaining or building up the topic outline, this should be a knowledge sharing topic outline and not include ongoing maintenance support

Author: done all

> ## What is Maximum Flow? — Concept Introduction

Reviewer: ```suggestion
### What is Maximum Flow? — Concept Introduction
```

Maximum Flow is the problem of sending as much "stuff" as possible from a source to a sink through a network of pipes (edges) that each have a capacity limit. The goal is to maximize the total amount leaving the source and arriving at the sink while never exceeding any pipe’s capacity.

Key terms:

> 

Reviewer: ```suggestion
```

- **Source (S)**: where flow starts.
- **Sink (T)**: where flow must end.
- **Capacity**: the maximum amount an edge can carry.
- **Flow**: how much is currently being sent through an edge (must be ≤ capacity).

Real-world analogies:

> 

Reviewer: ```suggestion
```

- Water pushed through pipes from a reservoir (source) to a treatment plant (sink).
- Trucks carrying packages through a roadway network.
- Data packets sent from a server to a client through routers with bandwidth limits.

> This page gives an intuitive, story-driven explanation, a small worked example, a visual residual-graph view, a beginner-friendly JavaScript implementation (Edmonds–Karp / BFS), complexity notes, guided and independent practice, and quick checks.

Reviewer: ```suggestion
```

Author: done all

> ## Interactive Narrative Scenario — The Bakery Delivery Network 🍞

Reviewer: ```suggestion
## Interactive Bakery Delivery Network Scenario 🍞
```

Imagine a bakery (`S`) that sends fresh bread to a city plaza (`T`) through a small distribution network of bike couriers and transfer hubs.

Nodes:

> 

Reviewer: ```suggestion
```

- `S` = Bakery (source)
- `A`, `B`, `C` = Hubs
- `T` = City Plaza (sink)

Capacities (loaves per hour) and a simple ASCII diagram:

Capacity table

| Edge  | Capacity |
| ----- | -------- |
| S → A | 10       |
| S → B | 5        |
| A → B | 15       |
| A → C | 4        |
| B → C | 8        |
| A → T | 5        |
| B → T | 10       |
| C → T | 15       |

ASCII diagram

S = source, T = sink

```
      (10)        (5)
  S --------> A ------> T
  |           | \      ^
  | (5)       |  \     |
  v           v   \    |
  B -----> C ------    |
   \ (10)     \ (15)  /
    \           \    /
     ----------->---
          (8)
```

- Source: `S`
- Sink: `T`
- We'll run the Ford–Fulkerson method (Edmonds–Karp variant using BFS) to find augmenting paths and push flow until no augmenting path exists.

## Step-by-Step Walkthrough (Augmenting Paths)

> 

Reviewer: ```suggestion
```

| A → C |        4 |    4 |        0 |
| B → C |        8 |    0 |        8 |
| A → T |        5 |    5 |        0 |
| B → T |       10 |    5 |        5 |
| C → T |       15 |    4 |       11 |

4. Augmenting path: S → A → B → T

- Note: S→A residual is 1; A→B residual is 15; B→T residual is 5
- Bottleneck = min(1, 15, 5) = 1
- Push 1 unit:
  - Flow S→A = 10 (now full)
  - Flow A→B = 1
  - Flow B→T = 6
- Total flow = 14 + 1 = 15

Updated table (after path 4 - final):

| Edge  | Capacity | Flow | Residual |
| ----- | -------: | ---: | -------: |
| S → A |       10 |   10 |        0 |
| S → B |        5 |    5 |        0 |
| A → B |       15 |    1 |       14 |
| A → C |        4 |    4 |        0 |
| B → C |        8 |    0 |        8 |
| A → T |        5 |    5 |        0 |
| B → T |       10 |    6 |        4 |
| C → T |       15 |    4 |       11 |

Now try to find another augmenting path from `S` to `T` in the residual graph: there is no path from `S` that reaches `T` with positive capacity (both S→A and S→B residuals are 0), so we stop.

Final maximum flow = 15 loaves/hour.

## Residual Graph & Backward Edges — Visual Explanation

Residual graph intuition:

> 

Reviewer: ```suggestion
```

- Every original edge u → v with capacity c and current flow f gives:
  - A **forward residual capacity** of c − f (how much more we can push forward).
  - A **backward residual capacity** of f (how much we could undo / push back).
- The residual graph contains both forward and backward edges. Augmenting paths are found in this residual graph.

Backward edge explanation:

> 

Reviewer: ```suggestion
```

- Backward edges let the algorithm "cancel" earlier choices. For example, if we later discover pushing some flow along an earlier path blocks a better combination, the algorithm can route flow backwards along an edge and reassign it elsewhere.

Why it works (intuitively):

> 

Reviewer: ```suggestion
```

- Each augmenting path strictly increases total flow by at least 1 unit (or more). We stop only when no path from S to T exists in the residual graph — at that point, you've saturated all S→T capacity possibilities.
- The Max-Flow Min-Cut Theorem: the maximum flow value equals the capacity of the minimum S–T cut (a partition of nodes separating S and T). When no augmenting path exists, the set of reachable nodes from S in the residual graph defines a cut whose capacity equals the current flow.

## JavaScript Implementation — Edmonds–Karp (BFS)

This implementation is beginner-friendly and logs each augmenting path, its bottleneck, and the running total flow. Paste into a file like `fordFulkerson.js` and run with `node fordFulkerson.js`.

```javascript
// fordFulkerson.js
// Edmonds-Karp (Ford-Fulkerson with BFS) implementation for teaching purposes.
//
// Graph represented as an adjacency matrix of capacities.
// We maintain a residual graph (capacities that can still be used).
//

> // The code logs each augmenting path, the bottleneck, and the updated total flow.

Reviewer: ```suggestion
/* fordFulkerson.js
Edmonds-Karp (Ford-Fulkerson with BFS) implementation
- Graph represented as an adjacency matrix of capacities.
- We maintain a residual graph (capacities that can still be used).
- The code logs each augmenting path, the bottleneck, and the updated total flow.
*/
```

      v = u;
    }
    path.push(source);
    path.reverse();

    // Log the augmenting path and bottleneck
    console.log('Augmenting path:', path.map((i) => nodeName(i)).join(' -> '));
    console.log('Bottleneck:', bottleneck);

    // Update residual capacities (forward and backward edges)
    v = sink;
    while (v !== source) {
      const u = parent[v];
      residual[u][v] -= bottleneck; // use capacity on forward edge
      residual[v][u] += bottleneck; // add capacity on backward edge
      v = u;
    }

    totalFlow += bottleneck;
    console.log('Total flow now:', totalFlow);
    console.log('---');
  }

  console.log('Max flow (final):', totalFlow);
  return totalFlow;
}

// Helper to map indices to node names for readable logs
function nodeName(i) {
  const labels = ['S', 'A', 'B', 'C', 'T'];
  return labels[i] || String(i);
}

// Example graph matched to the page example:
// node order: 0:S, 1:A, 2:B, 3:C, 4:T
const capacity = [
  //S  A   B   C   T
  [0, 10, 5, 0, 0], // S
  [0, 0, 15, 4, 5], // A
  [0, 0, 0, 8, 10], // B
  [0, 0, 0, 0, 15], // C
  [0, 0, 0, 0, 0] // T
];

// Run Edmonds-Karp
edmondsKarp(capacity, 0, 4);
```

What you should see when running:

> 

Reviewer: ```suggestion
```

Author: done all

- Lines logging each augmenting path found (e.g., `S -> A -> T`), its bottleneck (e.g., `5`), and the running total flow (e.g., `Total flow now: 5`).
- Final output `Max flow (final): 15`.

Notes for students:

> 

Reviewer: ```suggestion
```

- This code uses an adjacency matrix for clarity. For larger graphs, adjacency lists are more memory-efficient.
- The `residual` matrix keeps both forward and backward capacities so we can undo and reroute flow.

## Complexity Analysis

| Operation                              | Complexity                                                     |
| -------------------------------------- | -------------------------------------------------------------- |
| One BFS to find augmenting path        | O(V + E)                                                       |
| Number of augmentations (Edmonds–Karp) | O(V \* E)                                                      |
| Total time (Edmonds–Karp)              | O(V \* E^2) in worst-case bound; often much faster in practice |
| Space                                  | O(V^2) for adjacency matrix, O(V + E) for adjacency list       |

Explanation:

> 

Reviewer: ```suggestion
```

- Edmonds–Karp uses BFS to find the shortest augmenting path (by number of edges). Each BFS is O(V + E).
- Each augmentation increases the shortest-path distance of at least one edge on some path or saturates an edge; theoretical bounds give O(V \* E) augmentations in the worst case.
- Therefore total worst-case time is O(V \* E^2) when using adjacency matrix operations; with adjacency lists and smart implementations the practical performance is often much better.
- Using BFS avoids very deep/long augmenting sequences that naive DFS-based Ford–Fulkerson might fall into (which can be exponential for pathological cases).

## Guided Practice

Try this visualizer: https://cp-algorithms.com/graph/ford_fulkerson.html (contains runnable examples)  

> Or try an interactive web visualizer: https://visualgo.net/en/flow (enter the network and watch augmentations).

Reviewer: ```suggestion
- Try [this visualizer](https://cp-algorithms.com/graph/edmonds_karp.html) (contains runnable examples)  
- Or try an [interactive web visualizer] (https://visualgo.net/en/maxflow) (enter the network and watch augmentations).
```

Instructions for students:

> 

Reviewer: ```suggestion
Instructions:
```

- Open the visualizer and enter the example network (nodes `S, A, B, C, T`) and the capacities listed above.
- Run the Edmonds–Karp simulation step-by-step and observe:
  - Which augmenting path is chosen first (shortest path).
  - The bottleneck for each path.
  - How backward edges appear in the residual view.

Guided questions:

> 

Reviewer: ```suggestion
```

- Which edge became saturated first?
- After the second augmentation, which nodes are still reachable from `S` in the residual graph?

## Independent Practice

> 

Reviewer: ```suggestion
```

Author: done all

1. New problem: Modify the network by adding a new edge `S → C` with capacity 7 and recompute max flow. What changes in the augmenting sequence and final max flow?
2. Modify the JS code to:
   - Use an adjacency list instead of a matrix.
   - Track and print the actual flow per original edge in human-friendly form.
3. Identify the minimum cut for the original example. Which edges separate `S` from `T` when you partition reachable vs non-reachable nodes in the final residual graph?
4. Try swapping BFS for DFS in the code (naive Ford–Fulkerson). Does the number of augmentations change? Does runtime for larger random graphs change?

## Check-for-Understanding (5 Questions)

> 

Reviewer: ```suggestion
```

Author: done all

1. What is an augmenting path and why does it matter?
2. What is a bottleneck capacity on a path and how does it affect flow updates?
3. Why do we need backward edges in the residual graph? Give an example scenario where a backward edge is useful.
4. How does using BFS (Edmonds–Karp) help keep the algorithm efficient compared to arbitrary DFS paths?
5. Explain in plain language the Max-Flow Min-Cut Theorem and how it tells you that the algorithm has finished.

## Quick Tips for Contributors

- Keep the examples small and numerically easy for hand-tracing.
- Include a runnable JS snippet and sample output.
- Prefer BFS for teaching Edmonds–Karp; show DFS as a variant and explain pitfalls.
- When you add images or SVGs, include an accessible text version (ASCII or table) for screen readers.

---

> Created for the Techtonica curriculum.

Reviewer: ```suggestion
```
